File: features/travel/flight/src/main/java/com/tokopedia/flight/cancellation/view/adapter/FlightCancellationAdapterTypeFactory.java
package com.tokopedia.flight.cancellation.view.adapter;

import android.view.View;

import com.tokopedia.abstraction.base.view.adapter.factory.BaseAdapterTypeFactory;
import com.tokopedia.abstraction.base.view.adapter.viewholders.AbstractViewHolder;
import com.tokopedia.flight.cancellation.view.adapter.viewholder.FlightCancellationViewHolder;
import com.tokopedia.flight.cancellation.view.viewmodel.FlightCancellationModel;

/**
 * @author by furqan on 21/03/18.
 */

public class FlightCancellationAdapterTypeFactory extends BaseAdapterTypeFactory
        implements FlightCancellationTypeFactory {

    private FlightCancellationViewHolder.FlightCancellationListener flightCancellationListener;

    public FlightCancellationAdapterTypeFactory(FlightCancellationViewHolder.FlightCancellationListener flightCancellationListener) {
        this.flightCancellationListener = flightCancellationListener;
    }

    @Override
    public int type(FlightCancellationModel viewModel) {
        return FlightCancellationViewHolder.LAYOUT;
    }

    @Override
    public AbstractViewHolder createViewHolder(View parent, int type) {
        AbstractViewHolder viewHolder = null;
        if (type == FlightCancellationViewHolder.LAYOUT) {
            viewHolder = new FlightCancellationViewHolder(parent, flightCancellationListener);
        } else {
            viewHolder = super.createViewHolder(parent, type);
        }
        return viewHolder;
    }
}

File: features/travel/flight/src/main/java/com/tokopedia/flight/cancellation/presentation/adapter/FlightCancellationAdapterTypeFactory.kt
package com.tokopedia.flight.cancellation.presentation.adapter

import android.view.View
import com.tokopedia.abstraction.base.view.adapter.Visitable
import com.tokopedia.abstraction.base.view.adapter.factory.BaseAdapterTypeFactory
import com.tokopedia.abstraction.base.view.adapter.viewholders.AbstractViewHolder
import com.tokopedia.flight.cancellation.presentation.adapter.viewholder.FlightCancellationViewHolder
import com.tokopedia.flight.cancellation.presentation.model.FlightCancellationModel

/**
 * @author by furqan on 14/07/2020
 */
class FlightCancellationAdapterTypeFactory(private val flightCancellationListener: FlightCancellationViewHolder.FlightCancellationListener)
    : BaseAdapterTypeFactory(), FlightCancellationTypeFactory {

    override fun type(cancellationModel: FlightCancellationModel): Int = FlightCancellationViewHolder.LAYOUT

    override fun createViewHolder(parent: View, type: Int): AbstractViewHolder<out Visitable<*>> =
            when (type) {
                FlightCancellationViewHolder.LAYOUT -> FlightCancellationViewHolder(parent, flightCancellationListener)
                else -> super.createViewHolder(parent, type)
            }

}